machine/xen/v1alpha1.go

```diff
@@ -0,0 +1,536 @@
+		return machine, err
+	}
+
+	if len(machine.Status.LogFile) == 0 {
+		machine.Status.LogFile = filepath.Join(machine.Status.StateDir, "machine.log")
+	}
+
+	fd, err := os.Create(machine.Status.LogFile)
+	if err != nil {
+		return machine, err
+	}
+	defer fd.Close()
+
+	if machine.Spec.Resources.Requests == nil {
+		machine.Spec.Resources.Requests = make(corev1.ResourceList, 2)
+	}
+
+	if machine.Spec.Resources.Requests.Memory().Value() == 0 {
+		quantity, err := resource.ParseQuantity(fmt.Sprintf("%d", XenMemoryDefault) + "Mi")
+		if err != nil {
+			machine.Status.State = machinev1alpha1.MachineStateFailed
+			return machine, err
+		}
+
+		machine.Spec.Resources.Requests[corev1.ResourceMemory] = quantity
+	}
+
+	if machine.Spec.Resources.Requests.Cpu().Value() == 0 {
+		quantity, err := resource.ParseQuantity(fmt.Sprintf("%d", XenCPUsDefault))
+		if err != nil {
+			machine.Status.State = machinev1alpha1.MachineStateFailed
+			return machine, err
+		}
+
+		machine.Spec.Resources.Requests[corev1.ResourceCPU] = quantity
+	}
+
+	cfg.BInfo.MaxVcpus = int(machine.Spec.Resources.Requests.Cpu().Value())
+	cfg.BInfo.MaxMemkb = uint64(machine.Spec.Resources.Requests.Memory().Value() / XenMemoryScale)
+	cfg.BInfo.Kernel = machine.Status.KernelPath
+	cfg.CInfo.Name = string(machine.ObjectMeta.Name)
+	cfg.CInfo.Type = xenlight.DomainTypePv
+
+	if len(machine.Status.InitrdPath) > 0 {
+		cfg.BInfo.Ramdisk = machine.Status.InitrdPath
+	}
+
+	if len(machine.Spec.Ports) > 0 {
+		return machine, fmt.Errorf("mapping ports is not supported for xen")
+	}
+
+	kernelArgs, err := ukargparse.Parse(machine.Spec.KernelArgs...)
+	if err != nil {
+		return machine, err
+	}
+
+	if len(machine.Spec.Networks) > 0 {
+		startMac, err := macaddr.GenerateMacAddress(true)
+		if err != nil {
+			return machine, err
+		}
+
+		i := 0
+		for _, network := range machine.Spec.Networks {
+			for _, iface := range network.Interfaces {
+				mac := iface.Spec.MacAddress
+				if mac == "" {
+					startMac = macaddr.IncrementMacAddress(startMac)
+					mac = startMac.String()
+				}
+
+				nic, err := xenlight.NewDeviceNic()
+				if err != nil {
+					return nil, err
+				}
+
+				// TODO(andreistan26): Check if xen accepts CIDR notation
+				nic.Ip = fmt.Sprintf("%s %s %s", strings.Split(iface.Spec.CIDR, "/")[0], network.Netmask, network.Gateway)
+				nic.Bridge = network.IfName
+				nic.Mac = xenlight.Mac([]byte(mac))
+
+				cfg.Nics = append(cfg.Nics, *nic)
+				kernelArgs = append(kernelArgs,
+					uknetdev.NewParamIp().WithValue(uknetdev.NetdevIp{
+						CIDR:     iface.Spec.CIDR,
+						Gateway:  network.Gateway,
+						DNS0:     iface.Spec.DNS0,
+						DNS1:     iface.Spec.DNS1,
+						Hostname: iface.Spec.Hostname,
+						Domain:   iface.Spec.Domain,
+					}),
+				)
+				i++
+			}
+		}
+	}
+
+	var fstab []string
+
+	// TODO(andreistan26): Check if installed xen supports 9pfs
+	for i, vol := range machine.Spec.Volumes {
+		switch vol.Spec.Driver {
+		case "9pfs":
```

Comment:
not sure if value is from an `enum`. If not, then I would match also `9p` for now

Reply:
`vol.Spec.Driver` is a string, i've found no enum for `volume` types